### ISSUE TITLE
@sort errors when passing the name of a property as the second parameter.

### ISSUE THREAD

**Issue by NullVoxPopuli** [ember-decorators/ember-decorators#158]
This used to work:
```js
  sortBy: 'registeredAt:asc',
  @sort('model.registrations', 'sortBy') registrations: null,
```

but now it gives the error 
 `Assertion Failed: Cannot use 'sort' on field 'registrations' without a callback`

the only way to solve this is to use a full function for the comparator:

```js
  @sort('model.registrations', function(a, b) {
    const ra = a.get('registeredAt');
    const rb = b.get('registeredAt');

    return (ra > rb && 1) || (ra < rb && -1) || 0;
  }) registrations;
```

Fixed in PR: https://github.com/ember-decorators/ember-decorators/pull/157

**Comment by NullVoxPopuli** [comment]
Just noticed that the fix works with regular extends / object syntax, but not es6 classes :-(

**Comment by buschtoens** [comment]
Does https://github.com/ember-decorators/ember-decorators/pull/157#discussion_r134156959 work?

> Shouldn't that be `get(target, params[params.length - 1])`?

**Comment by NullVoxPopuli** [comment]
I saw that, and maybe I just misunderstood what get is doing?
I thought it was getting the value, which I don't really care about -- just that a value exists.

I tested it out though, and couldn't get it working. Is it just: `Ember.get(target, lastParam)`?

**Comment by buschtoens** [comment]
>  Is it just: `Ember.get(target, lastParam)`?

Yep.

> I saw that, and maybe I just misunderstood what get is doing?
> I thought it was getting the value, which I don't really care about -- just that a value exists.

True dat. But maybe `get` is doing some magic stuff that would make ES6 classes work. Evidently it's not. :smile:

**Comment by NullVoxPopuli** [comment]
@buschtoens something is fishy.

I have:
```js
console.log(target, lastParam, target.get(lastParam), Ember.get(target, lastParam));
```
which prints:
```
Class {constructor: ƒ} "sorts" undefined undefined
```

and my test is:
```js
  test('sort (no callback, use property value, es6)', function(assert) {
    class TestSort extends Ember.Object {
      constructor() {
        super();

        this.names = Ember.A([{name:'b'},{name:'z'},{name:'a'},{name:'foo'}]);
      }

      sorts = ['name:asc'];
      @sort('names', 'sorts') sortedNames;
    }

    var actual = (new TestSort).get('sortedNames').mapBy('name');
    assert.deepEqual(actual, ['a','b','foo','z']);
  });
```
this is all in https://github.com/ember-decorators/ember-decorators/pull/160

**Comment by buschtoens** [comment]
Yeah, that's what I meant by "Evidently it's not. " :wink: 

But looking at your code again, I think I know what's up: `sorts = ['name:asc'];` is transpiled to a `this.sorts =  ['name:asc'];` after the execution of the constructor. The decorator macro is probably executed *before* `sorts` is set.

Try setting `sorts` in the `init` hook before calling `this._super(...arguments);`.

**Comment by NullVoxPopuli** [comment]
adding `sorts` to the constructor ended up with the same error.
if this is a transpile order issue, I'm not sure we can assert on the  second parameter.  
Without changing anything, other than commenting out the assert, the test passes.

**Comment by pzuraq** [comment]
Yeah, the issue here is that ES Classes assign class properties per instance, whereas the old object model assigned them on the prototype of the class.

I think removing the assertion makes sense. It could cause some weird bugs if people `get` their sorted property in the `init/constructor` phase, but it also aligns with everyone having to learn this difference in how class properties will work. 

If users must get the sorted property early, they can either set it early (in the constructor), or use the `@proto` decorator we'll be adding to ember-decorators/utility-decorators to assign it directly to the prototype. Neither is particularly clean, but they'll both work.

**Comment by NullVoxPopuli** [comment]
@pzuraq so is that where the `static` notation comes in? would that define a property on the prototype as well?

**Comment by pzuraq** [comment]
`static` defines the property on the constructor, not the prototype, so unfortunately no 😕 currently there isn't a language level construct for defining properties on the prototype

<img width="813" alt="screen shot 2017-08-23 at 11 01 36 am" src="https://user-images.githubusercontent.com/685518/29630858-7c8cb93e-87f2-11e7-8438-350fe0d16beb.png">

**Comment by NullVoxPopuli** [comment]
sadness. 

@pzuraq have you seen https://github.com/ember-decorators/ember-decorators/pull/160 :-D

**Comment by pzuraq** [comment]
Yeah, I have, will post my thoughts there

**Comment by pzuraq** [comment]
Fixed by #160
